File: src/main/java/com/vc/easy/L687.scala
package com.vc.easy

class TreeNode687(var _value: Int) {
  var value: Int = _value
  var left: TreeNode687 = null
  var right: TreeNode687 = null
}

object L687 {
  def longestUnivaluePath(root: TreeNode687): Int = {
    var maxLength = 0
    def find(root: TreeNode687): Unit = {
      if(root != null) {
        val left = height(root.left, root.value)
        val right = height(root.right, root.value)
        maxLength = Math.max(maxLength, left + right)
        find(root.left)
        find(root.right)
      }
    }
    def height(root:TreeNode687, v:Int): Int = {
      if(root != null) {
        if(root.value == v) {
          return 1 + Math.max(height(root.left, v), height(root.right, v))
        }
      }
      0
    }
    find(root)
    maxLength
  }
}

File: src/main/java/com/vc/easy/L687.java
package com.vc.easy;

import com.vc.hard.TreeNode;

class L687 {
    private int max = 0;
    public int longestUnivaluePath(TreeNode root) {
        helper(root);
        return max;
    }

    private int helper(TreeNode root) {
        if(root == null) return 0;

        int left = helper(root.left);
        int right = helper(root.right);

        int leftValue = root.left == null || root.left.val != root.val ? 0 : left;
        int rightValue = root.right == null || root.right.val != root.val ? 0: right;

        max = Math.max(max, leftValue + rightValue);
        return 1 + Math.max(leftValue, rightValue);
    }
}
